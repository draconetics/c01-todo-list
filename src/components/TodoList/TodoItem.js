import React, { useState } from 'react';
import PropTypes from 'prop-types';

function TodoItem({ item, deleteTodo, updateTodo }) {
  const [todoContent, setTodoContent] = useState(item.todo || '');
  const [editMode, setEditMode] = useState(false);

  const handleChange = (e) => setTodoContent(e.target.value);

  const ItemText = () => {
    if (editMode) {
      return (
        <li>
          <input
            type="text"
            value={todoContent}
            onChange={(e) => handleChange(e)}
          />
        </li>
      );
    }
    return <li>{item.todo}</li>;
  };

  const changeToEditMode = () => setEditMode(true);
  const changeToShowMode = () => {
    setEditMode(false);
    setTodoContent(item.todo);
  };
  const updateItem = () => {
    updateTodo({ id: item.id, todo: todoContent });
    changeToShowMode();
  };

  const ItemOption = () => {
    if (editMode) {
      return (
        <li>
          <button type="button" onClick={() => updateItem()} className="btn">
            Save
          </button>
          <button type="button" onClick={() => changeToShowMode()} className="btn">
            Cancel
          </button>
        </li>
      );
    }
    return (
      <li>
        <button type="button" onClick={() => changeToEditMode()} className="btn">
          Edit
        </button>
        <button type="button" onClick={() => deleteTodo(item.id)} className="btn">
          Delete
        </button>
      </li>
    );
  };

  return (
    <ul data-test="TodoItem" className="todolist__content-row" key={item.id}>
      <li>{item.id}</li>
      {ItemText()}
      {ItemOption()}
    </ul>
  );
}

const item = {
  id: PropTypes.any,
  todo: PropTypes.string,
};
TodoItem.propTypes = {
  item: PropTypes.shape(item).isRequired,
  deleteTodo: PropTypes.func.isRequired,
  updateTodo: PropTypes.func.isRequired,
};

export default TodoItem;
