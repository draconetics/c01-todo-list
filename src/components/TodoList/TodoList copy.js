import React, {useRef, useState} from 'react';
import PropTypes from 'prop-types';

const TodoList = ({ todoList, addTodo, deleteTodo, editTodo })=>{

    const [todo, setTodo] = useState("");
    const refItems = useRef([]);
    const refEditItems = useRef([]);

    const insertTodo = (e)=>{
      e.preventDefault();
      console.log("inserting new element")
      const newTodo = {
        id: todoList.length +1,
        todo
      }
      addTodo({type:'ADD',value:newTodo});
      setTodo("");
    };
  
    const deleteTodoById = (id)=>{
        deleteTodo({type:'DELETE',value:{id}});
    }
  
    const editTodoByIndex = (index)=>{
      console.log(refItems.current[index]);
      refItems.current[index].classList.remove('show');
      refEditItems.current[index].classList.add('show');
    }
  
    const saveTodo =(id, index)=>{
        console.log(refEditItems.current[index].querySelector('input').value);
        let inputValue = refEditItems.current[index].querySelector('input').value;
        editTodo({type:'EDIT',value:{id,todo:inputValue}})
        cancelSetting(index);
    }
  
    const cancelSetting = (index)=>{
      refItems.current[index].classList.add('show');
      refEditItems.current[index].classList.remove('show');
    }

    return (
        <div className="todolist container" data-test="TodoList">
        <form onSubmit={(e)=>insertTodo(e)} className="todolist-add">
            <input type="text" value={todo} placeholder="I want to do..." onChange={(e)=>setTodo(e.target.value)}/>
            <button className="btn" type="submit">New Todo</button>
        </form>
        <div className="todolist-list">
              {
                    todoList && todoList.map((item,index)=>{
                        return (
                          <div key={index}>
                          <ul className="task-item show" ref={(element) => refItems.current.push(element)} >
                              <li>
                                {item.id}
                              </li>
                              <li>
                                  <div>{item.todo}</div>
                              </li>
                              <li>
                                <div>
                                    <button className="btn" onClick={()=>editTodoByIndex(index)}>edit</button>
                                    <button className="btn" onClick={()=>deleteTodoById(index)}>delete</button>
                                </div>
                              </li>
                          </ul>
                          <ul className="edit-task" ref={(element) => refEditItems.current.push(element)}>
                              <li>
                                {item.id}
                              </li>
                              <li>
                                  <input type="text" defaultValue={item.todo}></input>
                              </li>
                              <li>
                                <div>
                                    <button className="btn" onClick={()=>saveTodo(item.id, index)}>Save</button>
                                    <button className="btn" onClick={()=>cancelSetting(index)}>Cancel</button>
                                </div>
                              </li>
                          </ul>
                          
                          </div>
                          );
                    })
                }      
        </div>
          
      </div>
    );
}

const item = {
  id: PropTypes.any,
  todo: PropTypes.string,
};
TodoList.propTypes = {
  addTodo: PropTypes.func.isRequired,
  editTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  todoList: PropTypes.arrayOf(
    PropTypes.shape(item),
  ).isRequired,
};
export default TodoList;