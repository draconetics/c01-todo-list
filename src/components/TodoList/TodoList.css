.todolist{
    margin: auto;
    display: flex;
    flex-direction: column;
    text-align: center;
}
.todolist__form{
    width: 100%;
    padding: 2rem;
}

.todolist__form input {
    margin: 0 1rem;
}

.todolist__content{
    padding: 1rem;
}

.todolist__content-item, .todolist__content-edit-item{
    display: none;
}

.todolist__content-item.show, .todolist__content-edit-item.show{
    display: flex;
    justify-content: space-around;
    align-items: center;
    border-bottom: 1px solid rgba(0,0,0,.1);
    padding: 1rem;
}

