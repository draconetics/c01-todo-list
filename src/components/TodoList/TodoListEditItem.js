import React from 'react';
import PropTypes from 'prop-types';

export default function TodoListEditItem({item, index, saveTodo, cancelSetting, updateRefEditItems}) {
  return (
    <>
      <ul
        className="todolist__content-edit-item"
        ref={(element) => updateRefEditItems(index, element)}
      >
        <li>{item.id}</li>
        <li>
          <input type="text" defaultValue={item.todo}></input>
        </li>
        <li>
          <div>
            <button
              className="btn"
              onClick={() => saveTodo(item.id, index)}
            >
              Save
            </button>
            <button className="btn" onClick={() => cancelSetting(index)}>
              Cancel
            </button>
          </div>
        </li>
      </ul>
    </>
  );
}


TodoListEditItem.propTypes = {
    item: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    saveTodo: PropTypes.func.isRequired,
    cancelSetting: PropTypes.func.isRequired,
    updateRefEditItems: PropTypes.func.isRequired
}