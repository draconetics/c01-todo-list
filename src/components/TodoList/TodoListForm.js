import React from 'react';
import PropTypes from 'prop-types';

export default function TodoListForm({ todo, insertTodo, setTodoItem }) {
  return (
    <>
      <form onSubmit={(e) => insertTodo(e)} className="todolist__form">
        <input type="text" value={todo} placeholder="I want to do..." onChange={(e) => setTodoItem(e)} />
        <button className="btn" type="submit">New Todo</button>
      </form>
    </>
  );
}
TodoListForm.propTypes = {
  todo: PropTypes.string.isRequired,
  insertTodo: PropTypes.func.isRequired,
  setTodoItem: PropTypes.func.isRequired,
};
