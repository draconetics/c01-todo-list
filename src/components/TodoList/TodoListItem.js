import React from 'react';
import PropTypes from 'prop-types';

export default function TodoListItem({
  index,
  item,
  editTodoByIndex,
  deleteTodo,
  updateRefItems
}) {
  return (
    <>
      <ul
        className="todolist__content-item show"
        ref={(element) => updateRefItems(index, element)}
      >
        <li>{item.id}</li>
        <li>
          <div>{item.todo}</div>
        </li>
        <li>
          <div>
            <button className="btn" onClick={() => editTodoByIndex(index)}>
              edit
            </button>
            <button className="btn" onClick={() => deleteTodo(item.id, index)}>
              delete
            </button>
          </div>
        </li>
      </ul>
    </>
  );
}

TodoListItem.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.object,
  editTodoByIndex: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  updateRefItems: PropTypes.any.isRequired
};
