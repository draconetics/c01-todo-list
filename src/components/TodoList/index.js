import TodoList from './TodoList'
import {connect} from 'react-redux';

export function mapStateToProps(state) {
    return {
        todoList: state.todoReducer
    };
}

export const mapDispatchToProps = (dispatch)=>{
  
    return {
      addTodo: (newTodo) => dispatch({type:'ADD',value:newTodo}),
      deleteTodo: async (id) => await dispatch({type:'DELETE',value:{id}}),
      editTodo: (id, newTodo) => dispatch({type:'EDIT',value:{id,todo:newTodo}})
    };
  }
  
export default connect(mapStateToProps,mapDispatchToProps)(TodoList);