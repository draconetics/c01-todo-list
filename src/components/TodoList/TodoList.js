import React from 'react';
import PropTypes from 'prop-types';
import './TodoList.css';
import TodoListForm from './TodoListForm';
import TodoListItem from './TodoListItem';
import TodoListEditItem from './TodoListEditItem';

class TodoList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      todo: '',
    };
    this.refItems = [];
    this.refEditItems = [];
    this.updateReferences(this.props.todoList);
  }

  updateReferences = (todoList) => {
    if (
      todoList.length === this.refItems.length &&
      todoList === this.refEditItems.length
    )
      return;

    this.refItems = [todoList.length];
    this.refEditItems = [todoList.length];
    todoList.map((reg, index) => {
      this.refItems[index] = React.createRef();
      this.refEditItems[index] = React.createRef();
      return reg;
    });
  };

  insertRefTodo = () => {
    this.refItems.push(React.createRef());
    this.refEditItems.push(React.createRef());
  };

  insertTodo = (e) => {
    e.preventDefault();
    //console.log("inserting new element")
    //console.log(this.state);
    const { todoList, addTodo } = this.props;
    const newTodo = {
      id: todoList.length + 1,
      todo: this.state.todo,
    };
    addTodo(newTodo);
    //   this.insertRefTodo();
    this.setState({ todo: '' });
  };

  deleteTodo = (id, index) => {
    const { deleteTodo } = this.props;
    deleteTodo(id).then(() => {
      //this.deleteRefTodo(index);
    });
  };

  editTodoByIndex = (index) => {
    console.log('edit by reference');
    this.refItems[index].current.classList.remove('show');
    this.refEditItems[index].current.classList.add('show');
  };

  saveTodo = (id, index) => {
    const { editTodo } = this.props;
    let inputValue = this.refEditItems[index].current.querySelector('input')
      .value;
    editTodo(id, inputValue);
    this.cancelSetting(index);
  };

  cancelSetting = (index) => {
    this.refItems[index].current.classList.add('show');
    this.refEditItems[index].current.classList.remove('show');
  };

  setTodoItem = (e) => {
    console.log('setTodoItem');
    this.setState({ todo: e.target.value });
  };

  updateRefItems = (index, element) => {
    if (0 <= index && index < this.refItems.length)
      this.refItems[index].current = element;
  };

  updateRefEditItems = (index, element) => {
    if (0 <= index && index < this.refEditItems.length)
      this.refEditItems[index].current = element;
  };

  render = () => {
    const { todo } = this.state;
    const { todoList } = this.props;
    this.updateReferences(todoList);
    return (
      <div className="todolist container" data-test="TodoList">
        <TodoListForm
          todo={todo}
          insertTodo={this.insertTodo}
          setTodoItem={this.setTodoItem}
        />
        <div className="todolist__content">
          {todoList &&
            todoList.map((item, index) => {
              return (
                <div className="todolist__content-row" key={item.id}>
                  <TodoListItem
                    index={index}
                    item={item}
                    editTodoByIndex={this.editTodoByIndex}
                    deleteTodo={this.deleteTodo}
                    updateRefItems={this.updateRefItems}
                  />
                  <TodoListEditItem
                    item={item}
                    index={index}
                    saveTodo={this.saveTodo}
                    cancelSetting={this.cancelSetting}
                    updateRefEditItems={this.updateRefEditItems}
                  />
                </div>
              );
            })}
        </div>
      </div>
    );
  }; // end render
}

const item = {
  id: PropTypes.any,
  todo: PropTypes.string,
};
TodoList.propTypes = {
  addTodo: PropTypes.func.isRequired,
  editTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  todoList: PropTypes.arrayOf(PropTypes.shape(item)).isRequired,
};
export default TodoList;
