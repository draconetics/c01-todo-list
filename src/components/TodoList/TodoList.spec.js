import React from 'react';
import { mount, shallow } from 'enzyme';
import TodoList from './TodoList'

describe("#TodoList",()=>{
    describe("#TodoList Render", ()=>{
        
        let wrapper;

        beforeAll((done)=>{
            let props = {
                todoList:[], 
                addTodo:jest.fn(), 
                deleteTodo:jest.fn(), 
                editTodo:jest.fn()
            }
            wrapper = mount(<TodoList {...props}/>)
            done();
        })
        it('should render Menu without errors',()=>{
            const addTodoComponent = wrapper.find(`[data-test='TodoList']`);
            expect(addTodoComponent).toHaveLength(1);
        })
    })

    describe("#TodoList Events", ()=>{
        
        let wrapper;
        let spyAddTodo;

        beforeAll((done)=>{
            spyAddTodo = jest.fn();
            let props = {
                todoList:[{id:1,todo:"example01"},{id:2,todo:"example02"}], 
                addTodo:spyAddTodo, 
                deleteTodo:jest.fn(), 
                editTodo:jest.fn()
            }
            wrapper = mount(<TodoList {...props}/>)
            done();
        })
        it('should add new todo item', () => {
            const inputForm = wrapper.find('form input');
            expect(inputForm).toHaveLength(1);
            inputForm.simulate('change', { target: { value: 'new todo' } });

            const createButton = wrapper.findWhere((node) => node.type() === 'button' && node.text() === 'New Todo');
            expect(createButton).toHaveLength(1);
            createButton.simulate('click', {preventDefault: () => {}});
            
            console.log(wrapper.debug());
            
            expect(spyAddTodo).toHaveBeenCalledTimes(1);
            /* 
            expect(wrapper.find('.custom-modal h3').text()).toBe(
              'Create new register',
            ); */
          });
    })
})