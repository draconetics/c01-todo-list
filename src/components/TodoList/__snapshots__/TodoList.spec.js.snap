// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#TodoList #TodoList Render should render TodoList component without errors 1`] = `ReactWrapper {}`;
